src/tools/wasm-emscripten-finalize.cpp

```diff
@@ -76,28 +81,55 @@ int main(int argc, const char *argv[]) {
     WasmPrinter::printModule(&wasm, std::cerr);
   }
 
+  Export* dataEndExport = wasm.getExport("__data_end");
+  if (dataEndExport == nullptr) {
+    Fatal() << "__data_end export not found";
+  }
+  Global* dataEnd = wasm.getGlobal(dataEndExport->value);
+  if (dataEnd == nullptr) {
+    Fatal() << "__data_end global not found";
+  }
+  if (dataEnd->type != Type::i32) {
+    Fatal() << "__data_end global has wrong type";
+  }
+  Const* dataEndConst = dataEnd->init->cast<Const>();
+  uint32_t dataSize = dataEndConst->value.geti32() - globalBase;
+
+  std::vector<Name> initializerFunctions;
+  initializerFunctions.push_back("__wasm_call_ctors");
+
   EmscriptenGlueGenerator generator(wasm);
   generator.generateRuntimeFunctions();
   generator.generateMemoryGrowthFunction();
   generator.generateDynCallThunks();
-  generator.fixEmAsmConsts();
+  std::string metadata = generator.generateEmscriptenMetadata(dataSize, initializerFunctions);
 
   if (options.debug) {
     std::cerr << "Module after:\n";
     WasmPrinter::printModule(&wasm, std::cerr);
   }
 
-  ModuleWriter writer;
-  // writer.setDebug(options.debug);
-  writer.setDebugInfo(true);
-  // writer.setDebugInfo(options.passOptions.debugInfo);
-  // writer.setSymbolMap(symbolMap);
-  writer.setBinary(emitBinary);
-  // if (emitBinary) {
-  //   writer.setSourceMapFilename(sourceMapFilename);
-  //   writer.setSourceMapUrl(sourceMapUrl);
-  // }
-  writer.write(wasm, outfile);
+  if (!emitBinary) {
+    Output output(outfile, Flags::Text, Flags::Release);
+    WasmPrinter::printModule(&wasm, output.getStream());
```

Comment:
hmm, why does this manually do the text emitting? `ModuleWriter` can emit both text and binary.

Reply:
Because creating an `Output` object to handle the metadata wipes the file if we aren't printing to stdout. It always passes `std::ofstream::trunc`. We could get around this by:
A) adding a trunc/append flag, which is three sets of options to `Output`, which smells like we need a different factoring altogether.
B) modifying `ModuleWriter` to take an `Output` as a parameter to `write`, instead of / in addition to the filename APIs.
C) saying "screw it" and just doing it manually inline.

I went with C, but actually looking at the internals to `ModuleWriter`, it already uses `Output` internally, so it should be pretty trivial to change the interface to take `Output`s as parameters, and refactor the current interface as convenience wrappers around that. That sounds best.

Reply:
Done https://github.com/WebAssembly/binaryen/pull/1408/commits/50d8fa7b4979228aa9473ab7fcfd6920c03b02ad